### PR TITLE
Integrate spectra calculator for moist baro wave ex

### LINKED ISSUE #994: Integration of ClimaCoreSpectra in ClimaAtmos
# Purpose
This is an issue to track the integration work of a ClimaCore feature in ClimaAtmos. 
This issue is downstream of ClimaCore SDI #[967](https://github.com/CliMA/ClimaCore.jl/issues/967)

# Objectives
- [ ] Add spectrum calculations in the AMIP driver/postprocessing pipeline. Test case to be used for initial integration example is [moist baroclinic wave](https://github.com/CliMA/ClimaAtmos.jl/blob/c28ea7a16a581c72a4fd7783f2744c011a04b46b/examples/hybrid/define_post_processing.jl#L323)

# Components
- [ ] Implement a Kinetic Energy spectrum calculator in the `paperplots_moist_baro_wave_ρe` function. The intent is (at least initially) to replicate the existing behavior in [ClimateMachine](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Diagnostics/atmos_gcm_spectra.jl#L88), where we computed the spectrum of the `u` and `v` components separately and then computed the KE from these. 

## Producers
@valeriabarra 

## Reviewers/potential collaborators
Possibly @LenkaNovak  and @jiahe23

### PR DESCRIPTION
**Pull request by valeriabarra**
This PR will close #994 

## Purpose 
Integrate the use of `ClimaCoreSpectra` for spectra calculation in ClimaAtmos. 

Will close #994 .

## Content
Add spectrum calculations in the AMIP driver/postprocessing pipeline. The test case used for initial integration example is moist baroclinic wave. The intent is to replicate the existing behavior in [ClimateMachine](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Diagnostics/atmos_gcm_spectra.jl#L88) 
- Solution implemented as in the ClimateMachine implementation.

Review checklist

I have:
- [x] followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- [x] followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- [x] followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- [x] checked that this PR does not duplicate an open PR.

### CODE REVIEW

**Comment by valeriabarra** [comment]
bors try

File: examples/hybrid/define_post_processing.jl [hunk 1]
```diff
@@ -620,6 +621,49 @@ function paperplots_moist_baro_wave_ρe(sol, output_dir, p, nlat, nlon)
...
+            FT,
+            u,
+            similar(mass_weight),
+            lat,
+            lon,
+            mass_weight,
+        )
+        # Then compute 2d spectrum of u-component
+        u_2d_spectrum, wave_numbers, spherical, mesh_info =
+            power_spectrum_2d(FT, u[:, :, 1, 1], mass_weight) # use first level, 1500m
```
Comment (jiahe23) [inline examples/hybrid/define_post_processing.jl:637]: which is the index for the first level and what does the other `1` refer to?

File: examples/hybrid/define_post_processing.jl [hunk 2]
```diff
@@ -620,6 +621,49 @@ function paperplots_moist_baro_wave_ρe(sol, output_dir, p, nlat, nlon)
...
+        spectrum_1d = 0.5 .* (u_1d_spectrum + v_1d_spectrum)
+        spectrum_2d = 0.5 .* (u_2d_spectrum + v_2d_spectrum)
+        plot_spectrum = Plots.contourf(
+            collect(0:1:(mesh_info.num_fourier))[:],
+            collect(0:1:(mesh_info.num_spherical))[:],
+            (spectrum_2d[:, :, 1])',
+            xlabel = "m",
+            ylabel = "n",
+            color = :balance,
+            title = "2d spectrum (1500m) day " * string(day),
```
Comment (jiahe23) [inline examples/hybrid/define_post_processing.jl:660]: Does any vertical interpolation applied somewhere in the calculation? If not, in most of current set-ups, first layer may not correspond to 1500 meters.

File: examples/hybrid/define_post_processing.jl [hunk 3]
```diff
@@ -620,6 +621,49 @@ function paperplots_moist_baro_wave_ρe(sol, output_dir, p, nlat, nlon)
...
+        # Then compute 2d spectrum of v-component
+        v_2d_spectrum, wave_numbers, spherical, mesh_info =
+            power_spectrum_2d(FT, v[:, :, 1, 1], mass_weight) # use first level, 1500m
+
+        spectrum_1d = 0.5 .* (u_1d_spectrum + v_1d_spectrum)
+        spectrum_2d = 0.5 .* (u_2d_spectrum + v_2d_spectrum)
+        plot_spectrum = Plots.contourf(
+            collect(0:1:(mesh_info.num_fourier))[:],
+            collect(0:1:(mesh_info.num_spherical))[:],
+            (spectrum_2d[:, :, 1])',
```
Comment (jiahe23) [inline examples/hybrid/define_post_processing.jl:656]: what is the third dimension here?

File: examples/hybrid/define_post_processing.jl (hunk 2 above)
Comment (valeriabarra) [reply examples/hybrid/define_post_processing.jl:660]: Oh, I had referred to lines 574, 576 above where something similar was done for `u`. Is that different?

File: examples/hybrid/define_post_processing.jl (hunk 3 above)
Comment (valeriabarra) [reply examples/hybrid/define_post_processing.jl:656]: The dimensions of the variable `spectrum_2d` are `[m,n,k]`, where in this case there is only one level

File: examples/hybrid/define_post_processing.jl (hunk 1 above)
Comment (valeriabarra) [reply examples/hybrid/define_post_processing.jl:637]: Here `u` comes from the remapped data from the `NCDataset` above (lines 475 and unpacked on line 490). Its dimensions are `size(u) = (180, 90, 10, 1)` (for default CI pipeline resolution). So the third index is the vertical level, which for this spectrum calculation example I decided to use just the first one, and the fourth one I believe is just the component, so just 1 in this case.

File: examples/hybrid/define_post_processing.jl (hunk 2 above)
Comment (jiahe23) [reply examples/hybrid/define_post_processing.jl:660]: There might be a problem there as well ... This was originally hard-coded for the uniform grids of 10 layers with model top at 30km.

File: examples/hybrid/define_post_processing.jl [hunk 4]
```diff
@@ -620,6 +621,49 @@ function paperplots_moist_baro_wave_ρe(sol, output_dir, p, nlat, nlon)
...
+            xlabel = "m",
+            ylabel = "n",
+            color = :balance,
+            title = "2d spectrum (1500m) day " * string(day),
+        )
+        png(
+            plot_spectrum,
+            output_dir * "/bw-2dspectrum1500-day" * string(day) * ".png",
```
Comment (charleskawczynski) [inline examples/hybrid/define_post_processing.jl:664]:
```suggestion
            joinpath(output_dir, "bw-2dspectrum1500-day" * string(day) * ".png"),
```

File: examples/hybrid/define_post_processing.jl (hunk 2 above)
Comment (valeriabarra) [reply examples/hybrid/define_post_processing.jl:660]: If we want to fix all title strings, then I'd propose doing it in a different clean-up PR.

**Comment by valeriabarra** [comment]
bors try

**Comment by valeriabarra** [comment]
bors try

**Comment by valeriabarra** [comment]
@jiahe23 do you have any more questions/comments/review? Thanks

**Comment by jiahe23** [comment]
> @jiahe23 do you have any more questions/comments/review? Thanks

Could you add comments to the hard-coded index in case we forget again later? Otherwise it looks good to me.

BTW, is the latest CC working fine with hdf5 now?

**Comment by valeriabarra** [comment]
> Could you add comments to the hard-coded index in case we forget again later? Otherwise it looks good to me.

Done. 

> BTW, is the latest CC working fine with hdf5 now?
I am not aware of any issues but maybe I was out of the loop in that regard

**Comment by valeriabarra** [comment]
Branches are becoming stale really fast (I had to rebase onto `main` 3 times in the last half an hour), so if this is satisfactory, can I get an approval please so that we can merge it? Thanks

**Review by jiahe23 — APPROVED** [review]

**Review by charleskawczynski — COMMENTED** [review]
Ah, I forgot to hit submit review

**Pull request merged**